markup.js

```diff
@@ -164,83 +164,15 @@ class Markup {
   }
 
   static formatHTML (text = '', entities = []) {
-    const chars = [...text]
-    const available = [...entities]
-    const opened = []
-    const result = []
-    for (let offset = 0; offset < chars.length; offset++) {
-      while (true) {
-        const index = available.findIndex((entity) => entity.offset === offset)
-        if (index === -1) {
-          break
-        }
-        const entity = available[index]
-        switch (entity.type) {
-          case 'bold':
-            result.push('<b>')
-            break
-          case 'italic':
-            result.push('<i>')
-            break
-          case 'code':
-            result.push('<code>')
-            break
-          case 'pre':
-            result.push('<pre>')
-            break
-          case 'strikethrough':
-            result.push('<s>')
-            break
-          case 'underline':
-            result.push('<u>')
-            break
-          case 'text_mention':
-            result.push(`<a href="tg://user?id=${entity.user.id}">`)
-            break
-          case 'text_link':
-            result.push(`<a href="${entity.url}">`)
-            break
-        }
-        opened.unshift(entity)
-        available.splice(index, 1)
-      }
-
-      result.push(chars[offset])
-
-      while (true) {
-        const index = opened.findIndex((entity) => entity.offset + entity.length - 1 === offset)
-        if (index === -1) {
-          break
-        }
-        const entity = opened[index]
-        switch (entity.type) {
-          case 'bold':
-            result.push('</b>')
-            break
-          case 'italic':
-            result.push('</i>')
-            break
-          case 'code':
-            result.push('</code>')
-            break
-          case 'pre':
-            result.push('</pre>')
-            break
-          case 'strikethrough':
-            result.push('</s>')
-            break
-          case 'underline':
-            result.push('</u>')
-            break
-          case 'text_mention':
-          case 'text_link':
-            result.push('</a>')
-            break
-        }
-        opened.splice(index, 1)
-      }
-    }
-    return result.join('')
+    const chars = ['', ...text.split(''), '']
```

Comment:
Great code, however each character from `text` should be html-escaped:

```js
function escapeChar (c) {
  switch (c) {
    case `&`: return '&amp;'
    case `"`: return '&quot;'
    case `'`: return '&#39;'
    case `<`: return '&lt;'
    default : return c
  }
}
```

```suggestion
    const chars = ['', Array.prototype.map.call(text, escapeChar), '']
```

Reply:
Maybe `['', ...text.split(''), ''].map(escapeChar)`?